backend/src/main/java/gov/cdc/usds/simplereport/logging/PatientExperienceLoggingInterceptor.java

```diff
@@ -0,0 +1,74 @@
+package gov.cdc.usds.simplereport.logging;
+
+import gov.cdc.usds.simplereport.api.pxp.CurrentPatientContextHolder;
+import gov.cdc.usds.simplereport.db.model.Organization;
+import gov.cdc.usds.simplereport.db.model.PatientLink;
+import gov.cdc.usds.simplereport.service.AuditService;
+import java.util.UUID;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.slf4j.MDC;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnWebApplication;
+import org.springframework.stereotype.Component;
+import org.springframework.web.servlet.HandlerInterceptor;
+import org.springframework.web.servlet.ModelAndView;
+
+/**
+ * HandlerInterceptor to set a request ID for patient experience REST handlers, as well as logging
+ * audit information for successful patient requests. (Requests that raise exceptions are handled by
+ * {@link AuditLoggingAdvice#logAndRethrow(HttpServletRequest, Exception)}
+ */
+@Component
+@ConditionalOnWebApplication
+public class PatientExperienceLoggingInterceptor implements HandlerInterceptor {
+
+  private static final Logger LOG =
+      LoggerFactory.getLogger(PatientExperienceLoggingInterceptor.class);
+
+  @Autowired private CurrentPatientContextHolder _context;
+  @Autowired private AuditService _auditService;
+
+  @Override
+  public boolean preHandle(HttpServletRequest request, HttpServletResponse response, Object handler)
+      throws Exception {
+    LOG.trace(
+        "Pre-handling a method={} request for uri={} using handler={}",
+        request.getMethod(),
+        request.getRequestURI(),
+        handler);
+    String requestId = UUID.randomUUID().toString();
+    MDC.put(LoggingConstants.REQUEST_ID_MDC_KEY, requestId);
+    response.addHeader(LoggingConstants.REQUEST_ID_HEADER, requestId);
+    return true;
+  }
+
+  @Override
+  public void postHandle(
+      HttpServletRequest request,
+      HttpServletResponse response,
+      Object handler,
+      ModelAndView modelAndView)
+      throws Exception {
+    PatientLink patientLink = _context.getPatientLink();
+    LOG.trace("Closing out request. Patient link is {}", patientLink);
+    int responseCode = response.getStatus();
+    if (patientLink != null) {
+      String requestId = MDC.get(LoggingConstants.REQUEST_ID_MDC_KEY);
+      Organization organization = _context.getOrganization();
+      _auditService.logRestEvent(requestId, request, responseCode, organization, patientLink);
+    } else {
+      LOG.trace("No patient link found"); // unauthenticated handlers
+    }
+  }
+
+  @Override
+  public void afterCompletion(
+      HttpServletRequest request, HttpServletResponse response, Object handler, Exception ex)
+      throws Exception {
+    LOG.trace("Final logging cleanup step.");
+    MDC.remove(LoggingConstants.REQUEST_ID_MDC_KEY);
+  }
```

Comment:
Spring n00b question: if logging were happening here instead of in `postHandle`, would you need a separate `ControllerAdvice` logger to intercept exceptions? Not sure if there are some conditions under which Spring would not invoke `onComplete`.

Reply:
Not a bad question at all, actually—I spent a chunk of the weekend trying to get a good answer on it. To the best of my belief, you would not need a separate ControllerAdvice _but_ you would also not be able to fully enforce the AU-5 restriction on responses going out without logging. By the time we get here, the response is committed, so we can't change the content if the audit log fails. Since I am operating under the assumption that I can't hand-wave that away, I went with the ControllerAdvice.

I could of course be totally wrong in that belief—I didn't experiment!—but I did research it at least a bit, and this is where I landed.

Reply:
LOL turns out I was right about it not working in afterCompletion, but wrong about it working in postHandle. &lt;sigh&gt;